### PR TITLE
feat: add compute_makespan and test

### PR DESCRIPTION
**Pull request by MarcBoelle3**
Given the adjacency matrix of the graph (machine constraints + precedence constraints), computes the makespan of the schedule, which is used for the reward. It also computes, for each operation, the earliest and latest starting times. This allows to identify operations located on the critical path.

### CODE REVIEW

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    received_c = received_attributes[:, 1]
+
+    return jnp.stack([node_p, received_d, received_c], axis=-1)
+
+
+def forward_pass_jraph(
+    adj_mat: chex.Array, ops_durations: chex.Array, max_num_edges: jnp.int32
+) -> Tuple[chex.Array, jnp.float32]:
+    """Compute earliest start times and makespan using jraph message passing with max pooling.
+        First part of the algorithm in Section 4.4 of the paper.
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/compute_makespan.py:68]: I would add a link to the paper here so people can check out the algorithm easily

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    Args:
+        nodes: Array of nodes with shape (num_nodes, 3)
+        _sent_attributes: Array of sent attributes with shape (num_edges, 2)
+        received_attributes: Array of received attributes with shape (num_edges, 2)
+        _globals: Array of globals with shape (num_globals, 1)
+    Returns:
+        Array of updated nodes with shape (num_nodes, 3)
+    """
+
+    # Each node has a feature (p, d, c) with p the processing time,
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/compute_makespan.py:53]: Can we give better names than p,d,c maybe ?
I think we have that in the state too, but here the computations are complex, i think it would be really helpful to help people understand whats happening here as much as possible

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py (hunk 1 above)
Comment (mintiti) [reply jumanji/environments/packing/job_shop/improvement/compute_makespan.py:68]: Maybe explain the idea behind the algorithm too

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    node_p = nodes[:, 0]  # remains unchanged
+
+    # Update date and completion status if neighbours
+    received_d = received_attributes[:, 0]
+    received_c = received_attributes[:, 1]
+
+    return jnp.stack([node_p, received_d, received_c], axis=-1)
+
+
+def forward_pass_jraph(
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/compute_makespan.py:64]: Since this computes the makespan, does it make sense to put it in the name ? 
It would help to understand why and where you should use this function

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+import chex
+import jax
+import jax.numpy as jnp
+import jraph
+
+
+def update_node_fn(
+    nodes: chex.Array,
+    _sent_attributes: chex.Array,
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/compute_makespan.py:38]: Whats the reasoning for having this and _globals as private ?

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+    # Return earliest start times and makespan
+    return output_graph.nodes[:, 1], output_graph.nodes[-1, 1]
+
+
+def backward_pass_jraph(
+    adj_mat: chex.Array, ops_durations: chex.Array, makespan: jnp.float32, max_num_edges: jnp.int32
+) -> chex.Array:
+    """Compute latest start times using message passing with max pooling.
+        Second part of the algorithm in Section 4.4 of the paper.
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/compute_makespan.py:186]: Same thing here, maybe add a quick link in MarkDown format

File: jumanji/environments/packing/job_shop/improvement/compute_makespan_test.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        """
+        Example matrices for the forward and backward pass.
+        There are 8 valid operations, from 0 to 7. Machines are given by ops_machines_ids.
+        adj_mat_mc corresponds to the following operation order on machines:
+        - Machine 0: 0 -> 3
+        - Machine 1: 1 -> 6 -> 5
+        - Machine 2: 4 -> 2 -> 7
+        """
+
+        # For indication
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/compute_makespan_test.py:45]: Should we make this a comment if the idea is just to show it maybe?

File: jumanji/environments/packing/job_shop/improvement/compute_makespan_test.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        lst = backward_pass_jraph(adj_mat, ops_durations, makespan, MAX_NUM_EDGES)
+
+        # Verify source node has lst = 0 and last node has lst = makespan
+        assert lst[0] == 0
+        assert lst[-1] == 13
+
+        # Verify each operation has correct latest start time
+        assert jnp.all(lst[1:9] == jnp.array([0, 3, 8, 5, 7, 9, 5, 10]))
+
+    def test_jit_backward_pass_jraph(
```
Comment (mintiti) [inline jumanji/environments/packing/job_shop/improvement/compute_makespan_test.py:180]: And is it doable to have the same tests on the backward pass ? I think you dont test the simple chain for the backwards pass for example

**Review by mintiti — APPROVED** [review]
Looks great, good job !
Just a few changes to help make things easier to understand, especially because the calculations are far from trivial.
Resolve it and you can merge

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py (hunk 4 above)
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/compute_makespan.py:38]: it is to underline that it is not used in the computation, but is necessary for jraph framework

File: jumanji/environments/packing/job_shop/improvement/compute_makespan.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    """
+    num_nodes = adj_mat.shape[0]
+
+    # convert ops_durations to a vector of size num_nodes
+    ops_durations = ops_durations.reshape(-1)
+    # add sink and target nodes with duration of 0
+    ops_durations = jnp.concatenate([jnp.array([0.0]), ops_durations, jnp.array([0.0])])
+
+    # Create node features: (processing time, date, completion status)
+    node_features = jnp.ones((num_nodes, 3))
```
[the remark this answers is not in the document]
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/compute_makespan.py:205]: I think that when using vmap, provided that the graphs have the same shape (which is the case here thanks to static shapes), jraph handles automatically batched inputs...

[the remark this answers is not in the document]
Comment (MarcBoelle3) [reply jumanji/environments/packing/job_shop/improvement/compute_makespan.py:205]: but agree I should check how it is done in depth to maybe have better performance
